integration/core/test_identity.py

```diff
@@ -103,7 +103,7 @@ def test_validation_fails_with_client(engine_manager_client,
     with pytest.raises(grpc.RpcError) as e:
         e_client.replica_create('tcp://' +
                                 get_process_address(replica_different))
-    assert e.value.code() == grpc.StatusCode.UNKNOWN
+    assert e.value.code() == grpc.StatusCode.FAILED_PRECONDITION
```

Comment:
Why did the returned status change? Does this imply that there have been recent changes to the returned status of these methods?

The same questions to other status assertion change.

Reply:
The `FAILED_PRECONDITION` and `incorrect volume name` error message (in line 107) were introduced by https://github.com/longhorn/longhorn-engine/pull/902. The behavior change maybe cause of updating python grpcio from `1.25.0` to `1.58.0`. @ejweber, could you help to check whether status code `FAILED_PRECONDITION` is reasonable here? Thank you.

Reply:
I see, but why did these cases not fail recently in PR checks before this change?

Reply:
I can't find any logic change in our code. The different part may be `serialized_options` in generated files:

* Old file: https://github.com/longhorn/longhorn-engine/blob/8469537f32a03e1f8b1fdfbb18c8a374d2785743/integration/rpc/controller/controller_pb2.py#L222
* PR change: https://github.com/FrankYang0529/longhorn-engine/blob/075afeeb813ab8a9798b4455ad099b63dc440a18/integration/rpc/controller/github/com/longhorn/longhorn_engine/proto/ptypes/controller_pb2.py#L28

Reply:
I'm not able to understand this yet.

The replica service returns FAILED_PRECONDITION to the controller service, but IIRC, the controller service doesn't automatically return the same code to the controller client. Since it doesn't recognize the error in particular, it just returns an UNKNOWN. That is why the original test checked for UNKNOWN and also the specific text on line 107.

Maybe some change now allows the FAILED_PRECONDITION to propagate from replica service -> controller service -> controller client?

Reply:
Will these cases fail w/o these changes? If yes, we need to figure out the reasons.

Reply:
After checking the code, the assertion change is unnecessary here. What you encountered should be somehow the call did not take effect due to service problems. One example in the current test cases is below where we hit the failed precondition error when the client is unable to reach out to the server. 

[/integration/core/test_identity.py#L80-L88](https://github.com/longhorn/longhorn-engine/blob/2e09ee964d970ab6cbd26b40d942e7b1e9d9a9f6/integration/core/test_identity.py#L80-L88)
```py
    # CASE 6:
    # Our sync agent client is created with a different instance name than the
    # sync agent.
    # We cannot communicate with the sync agent.
    s_client = SyncAgentClient(get_sync_agent_service_address(replica),
                               VOLUME_NAME, 'wrong')
    with pytest.raises(grpc.RpcError) as e:
        s_client.replica_rebuild_status()
    assert e.value.code() == grpc.StatusCode.FAILED_PRECONDITION
```

Just checked the current the engine gRPC implementation, which did not handle any specific gRPC status error when returning an error from a method call. In this case, gRPC will just return an unknown status code.

https://github.com/grpc/grpc-go/blob/v1.58.2/status/status.go#L96-L127

Suggest rerunning the test cases to figure out why the tests failed instead.

Reply:
My fault. The change is necessary, as the current engine/replica gRPC server has been implemented with an interceptor which will issue `FAILED_PRECONDITION` if the engine can't communicate with the replica. 

[/proto/ptypes/interceptor.go#L54-L58](https://github.com/longhorn/longhorn-engine/blob/5464ea87c09fb0dd2aed575127b1d7c3756cfd12/proto/ptypes/interceptor.go#L54-L58)
```go
				if incomingInstanceName != instanceName {
					log.Error("Invalid gRPC metadata")
					return nil, status.Errorf(codes.FailedPrecondition, "incorrect instance name %s; check %s address",
						incomingInstanceName, serverType)
				}
```

Reply:
cc @ejweber